platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -13,204 +13,259 @@
+    if (fileFromCache == null && fileFromConfig == null) {
+      // This config doesn't have an etc file, so we ignore this case (1)
       LOGGER.debug("Was not tracked and will not track pid {}", pid);
       return;
     }
 
-    boolean filePropChanged = felixConfig.filePropChanged(felixFileOriginal);
-
-    // Felix file prop did not change? (4)
-    if (filePropChanged) {
-      // Are we seeing the config for the first time? (2) (3)
-      if (felixFileOriginal != null) {
-        try {
-          felixConfig.setFelixFile(felixFileOriginal);
-        } catch (IOException e) {
-          LOGGER.error("Could not set felix file name, error writing to config admin: ", e);
-        }
-        return;
-      }
-      pidFileMap.put(pid, felixFileNew);
+    if (fileFromCache == null) {
+      // An etc config file was just dropped and we're seeing it for the first time (2)
       LOGGER.debug("Tracking pid {}", pid);
+      CachedConfigData createdConfigData = new CachedConfigData(context);
+      pidDataMap.put(pid, createdConfigData);
+      writeIfNecessary(context, createdConfigData);
+      return;
+    }
+
+    if (!Objects.equals(fileFromConfig, fileFromCache)) {
+      // The felix file prop changed, which is not allowed (3)
+      String msg =
+          (fileFromConfig == null)
+              ? format("Felix filename has been illegally removed, was [%s]", fileFromCache)
+              : format(
+                  "Felix filename has been illegally changed from [%s] to [%s]",
+                  fileFromCache, fileFromConfig);
+      throw new IllegalStateException(msg);
     }
+
+    // Write to disk if necessary (4)
+    writeIfNecessary(context, cachedConfigData);
   }
 
   /**
-   * Remove config files from etc when a delete occurs but only if felix is tracking the file.
+   * {@inheritDoc}
+   *
+   * <p>Remove config files from etc when a delete occurs but only if felix is tracking the file.
    *
-   * @param event the configuration event for which a delete is needed
+   * @param pid the pid pointing to the configuration to delete.
    */
-  private void handleDelete(ConfigurationEvent event) {
-    String pid = event.getPid();
-    File felixFileOriginal = pidFileMap.remove(pid);
-    if (felixFileOriginal != null) {
+  @Override
+  public void handleDelete(String pid) throws IOException {
+    CachedConfigData cachedConfigData = pidDataMap.remove(pid);
+    File fileFromCache = (cachedConfigData != null) ? cachedConfigData.getFelixFile() : null;
+    if (fileFromCache != null && fileFromCache.exists()) {
       LOGGER.debug("Deleting file because config was deleted for pid {}", pid);
       try {
-        performOperation(configActions.delete(felixFileOriginal.toPath()));
-      } catch (IllegalArgumentException | ConfiguratorException e) {
+        Files.delete(fileFromCache.toPath());
+        SECURITY_LOGGER.info("Removing a deleted config: {}", fileFromCache.toPath());
```

Comment:
It does seem that this would be audited higher up. @stustison do you think this is necessary here?

Reply:
I recall our policy being something along the lines of **all** alterations to disk need to be audited. I'm not clear if that's the same for reads, but I think it might be.

Reply:
My point here is that if we were not building this component, Felix would be updating and deleting files on disk for us already and there would be no auditing there. This is a core component of the OSGI implementation we have. I don't think we should start auditing files being updated/deleted here or in the cache for that matter. But that is just me. I think @stustison should provide insight here.